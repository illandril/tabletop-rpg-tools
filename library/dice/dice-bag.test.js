/* SPDX-License-Identifier: MIT */

import shortid from 'shortid';

import RNGMulberry32 from './random-number-generator-mulberry32.js';
import DiceBag from './dice-bag.js';

beforeEach(() => {
  let nextShortID = 1;
  jest.spyOn(shortid, 'generate').mockImplementation(() => 'MOCK-' + nextShortID++);
});

afterEach(() => {
  shortid.generate.mockRestore();
});

test('default parser and rng', () => {
  // Make sure the default parser works and uses Math.random
  const diceBag = new DiceBag();
  try {
    for (let i = 0; i < 12; i++) {
      jest.spyOn(Math, 'random').mockReturnValue(i / 12);
      expect(diceBag.rollValue('1d6')).toEqual(Math.ceil((i + 1) / 2));
      expect(diceBag.rollValue('1d12')).toEqual(i + 1);
    }
  } finally {
    Math.random.mockRestore();
  }
  expect(diceBag.rollValue('125/25*3+16/8-33/11+10/5')).toEqual(16);
});

test('invalid arguments', () => {
  // prettier-ignore
  const invalidRNGOrParserMessage = 'DiceBag accepts only a random number generator with a d function, or a parser with a parse function';
  expect(() => new DiceBag('pastry')).toThrowError(invalidRNGOrParserMessage);
  expect(() => new DiceBag({})).toThrowError(invalidRNGOrParserMessage);
  expect(() => new DiceBag({ d: 5 })).toThrowError(invalidRNGOrParserMessage);
  expect(() => new DiceBag({ parse: '17' })).toThrowError(invalidRNGOrParserMessage);
});

test('custom rng', () => {
  const dValues = [1, 2, 3, 4, 5, 6, 1, 20];
  const mockD = jest.fn(() => dValues.shift());
  const rng = {
    d: mockD,
  };
  const diceBag = new DiceBag(rng);
  expect(mockD.mock.calls.length).toEqual(0);
  expect(diceBag.rollValue('6d6kl2')).toEqual(3);
  expect(mockD.mock.calls.length).toEqual(6);
  expect(mockD.mock.calls[0][0]).toEqual(6);
  expect(mockD.mock.calls[1][0]).toEqual(6);
  expect(mockD.mock.calls[2][0]).toEqual(6);
  expect(mockD.mock.calls[3][0]).toEqual(6);
  expect(mockD.mock.calls[4][0]).toEqual(6);
  expect(mockD.mock.calls[5][0]).toEqual(6);
  mockD.mockClear();

  expect(diceBag.rollValue('125/25*3+16/8-33/11+10/5')).toEqual(16);
  expect(mockD.mock.calls.length).toEqual(0);
  expect(diceBag.rollValue('1d20')).toEqual(1);
  expect(diceBag.rollValue('1d20')).toEqual(20);
  expect(mockD.mock.calls.length).toEqual(2);
  expect(mockD.mock.calls[0][0]).toEqual(20);
  expect(mockD.mock.calls[1][0]).toEqual(20);
  mockD.mockClear();
});

test('custom parser', () => {
  const parseValues = [{ value: 1 }, { value: 8 }];
  const mockParse = jest.fn(() => parseValues.shift());
  const parser = {
    parse: mockParse,
  };
  const diceBag = new DiceBag(parser);
  expect(mockParse.mock.calls.length).toEqual(0);
  expect(diceBag.rollValue('fishsticks')).toEqual(1);
  expect(mockParse.mock.calls.length).toEqual(1);
  expect(mockParse.mock.calls[0][0]).toEqual('fishsticks');
  mockParse.mockClear();

  expect(diceBag.rollValue('2d20kl1')).toEqual(8);
  expect(mockParse.mock.calls.length).toEqual(1);
  expect(mockParse.mock.calls[0][0]).toEqual('2d20kl1');
  mockParse.mockClear();
});

test('rollValue', () => {
  const rng = new RNGMulberry32(0);
  const diceBag = new DiceBag(rng);
  [1, 2, 4].forEach((count) => {
    [6, 10, 20].forEach((sides) => {
      for (let i = 0; i < 5; i++) {
        const dn = count + 'd' + sides;
        expect(diceBag.rollValue(dn)).toMatchSnapshot(dn);
      }
    });
  });
});

test('r', () => {
  const rng = new RNGMulberry32(0);
  const diceBag = new DiceBag(rng);
  [1, 2, 4].forEach((count) => {
    [6, 10, 20].forEach((sides) => {
      for (let i = 0; i < 5; i++) {
        const dn = count + 'd' + sides;
        expect(diceBag.rollValue(dn)).toMatchSnapshot(dn);
      }
    });
  });
});

test('roll', () => {
  const rng = new RNGMulberry32(0);
  const diceBag = new DiceBag(rng);
  [1, 2, 4].forEach((count) => {
    [6, 10, 20].forEach((sides) => {
      for (let i = 0; i < 5; i++) {
        const dn = count + 'd' + sides;
        expect(diceBag.rollValue(dn)).toMatchSnapshot(dn);
      }
    });
  });
});
