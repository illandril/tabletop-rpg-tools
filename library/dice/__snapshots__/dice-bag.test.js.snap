// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`r: 1d6 1`] = `2`;

exports[`r: 1d6 2`] = `1`;

exports[`r: 1d6 3`] = `2`;

exports[`r: 1d6 4`] = `1`;

exports[`r: 1d6 5`] = `3`;

exports[`r: 1d10 1`] = `6`;

exports[`r: 1d10 2`] = `7`;

exports[`r: 1d10 3`] = `7`;

exports[`r: 1d10 4`] = `5`;

exports[`r: 1d10 5`] = `6`;

exports[`r: 1d20 1`] = `5`;

exports[`r: 1d20 2`] = `2`;

exports[`r: 1d20 3`] = `19`;

exports[`r: 1d20 4`] = `9`;

exports[`r: 1d20 5`] = `15`;

exports[`r: 2d6 1`] = `9`;

exports[`r: 2d6 2`] = `8`;

exports[`r: 2d6 3`] = `3`;

exports[`r: 2d6 4`] = `7`;

exports[`r: 2d6 5`] = `6`;

exports[`r: 2d10 1`] = `16`;

exports[`r: 2d10 2`] = `9`;

exports[`r: 2d10 3`] = `9`;

exports[`r: 2d10 4`] = `9`;

exports[`r: 2d10 5`] = `6`;

exports[`r: 2d20 1`] = `13`;

exports[`r: 2d20 2`] = `10`;

exports[`r: 2d20 3`] = `24`;

exports[`r: 2d20 4`] = `19`;

exports[`r: 2d20 5`] = `31`;

exports[`r: 4d6 1`] = `17`;

exports[`r: 4d6 2`] = `10`;

exports[`r: 4d6 3`] = `15`;

exports[`r: 4d6 4`] = `5`;

exports[`r: 4d6 5`] = `16`;

exports[`r: 4d10 1`] = `24`;

exports[`r: 4d10 2`] = `28`;

exports[`r: 4d10 3`] = `13`;

exports[`r: 4d10 4`] = `24`;

exports[`r: 4d10 5`] = `24`;

exports[`r: 4d20 1`] = `35`;

exports[`r: 4d20 2`] = `31`;

exports[`r: 4d20 3`] = `42`;

exports[`r: 4d20 4`] = `45`;

exports[`r: 4d20 5`] = `37`;

exports[`roll: 1d6 1`] = `2`;

exports[`roll: 1d6 2`] = `1`;

exports[`roll: 1d6 3`] = `2`;

exports[`roll: 1d6 4`] = `1`;

exports[`roll: 1d6 5`] = `3`;

exports[`roll: 1d10 1`] = `6`;

exports[`roll: 1d10 2`] = `7`;

exports[`roll: 1d10 3`] = `7`;

exports[`roll: 1d10 4`] = `5`;

exports[`roll: 1d10 5`] = `6`;

exports[`roll: 1d20 1`] = `5`;

exports[`roll: 1d20 2`] = `2`;

exports[`roll: 1d20 3`] = `19`;

exports[`roll: 1d20 4`] = `9`;

exports[`roll: 1d20 5`] = `15`;

exports[`roll: 2d6 1`] = `9`;

exports[`roll: 2d6 2`] = `8`;

exports[`roll: 2d6 3`] = `3`;

exports[`roll: 2d6 4`] = `7`;

exports[`roll: 2d6 5`] = `6`;

exports[`roll: 2d10 1`] = `16`;

exports[`roll: 2d10 2`] = `9`;

exports[`roll: 2d10 3`] = `9`;

exports[`roll: 2d10 4`] = `9`;

exports[`roll: 2d10 5`] = `6`;

exports[`roll: 2d20 1`] = `13`;

exports[`roll: 2d20 2`] = `10`;

exports[`roll: 2d20 3`] = `24`;

exports[`roll: 2d20 4`] = `19`;

exports[`roll: 2d20 5`] = `31`;

exports[`roll: 4d6 1`] = `17`;

exports[`roll: 4d6 2`] = `10`;

exports[`roll: 4d6 3`] = `15`;

exports[`roll: 4d6 4`] = `5`;

exports[`roll: 4d6 5`] = `16`;

exports[`roll: 4d10 1`] = `24`;

exports[`roll: 4d10 2`] = `28`;

exports[`roll: 4d10 3`] = `13`;

exports[`roll: 4d10 4`] = `24`;

exports[`roll: 4d10 5`] = `24`;

exports[`roll: 4d20 1`] = `35`;

exports[`roll: 4d20 2`] = `31`;

exports[`roll: 4d20 3`] = `42`;

exports[`roll: 4d20 4`] = `45`;

exports[`roll: 4d20 5`] = `37`;

exports[`rollValue: 1d6 1`] = `2`;

exports[`rollValue: 1d6 2`] = `1`;

exports[`rollValue: 1d6 3`] = `2`;

exports[`rollValue: 1d6 4`] = `1`;

exports[`rollValue: 1d6 5`] = `3`;

exports[`rollValue: 1d10 1`] = `6`;

exports[`rollValue: 1d10 2`] = `7`;

exports[`rollValue: 1d10 3`] = `7`;

exports[`rollValue: 1d10 4`] = `5`;

exports[`rollValue: 1d10 5`] = `6`;

exports[`rollValue: 1d20 1`] = `5`;

exports[`rollValue: 1d20 2`] = `2`;

exports[`rollValue: 1d20 3`] = `19`;

exports[`rollValue: 1d20 4`] = `9`;

exports[`rollValue: 1d20 5`] = `15`;

exports[`rollValue: 2d6 1`] = `9`;

exports[`rollValue: 2d6 2`] = `8`;

exports[`rollValue: 2d6 3`] = `3`;

exports[`rollValue: 2d6 4`] = `7`;

exports[`rollValue: 2d6 5`] = `6`;

exports[`rollValue: 2d10 1`] = `16`;

exports[`rollValue: 2d10 2`] = `9`;

exports[`rollValue: 2d10 3`] = `9`;

exports[`rollValue: 2d10 4`] = `9`;

exports[`rollValue: 2d10 5`] = `6`;

exports[`rollValue: 2d20 1`] = `13`;

exports[`rollValue: 2d20 2`] = `10`;

exports[`rollValue: 2d20 3`] = `24`;

exports[`rollValue: 2d20 4`] = `19`;

exports[`rollValue: 2d20 5`] = `31`;

exports[`rollValue: 4d6 1`] = `17`;

exports[`rollValue: 4d6 2`] = `10`;

exports[`rollValue: 4d6 3`] = `15`;

exports[`rollValue: 4d6 4`] = `5`;

exports[`rollValue: 4d6 5`] = `16`;

exports[`rollValue: 4d10 1`] = `24`;

exports[`rollValue: 4d10 2`] = `28`;

exports[`rollValue: 4d10 3`] = `13`;

exports[`rollValue: 4d10 4`] = `24`;

exports[`rollValue: 4d10 5`] = `24`;

exports[`rollValue: 4d20 1`] = `35`;

exports[`rollValue: 4d20 2`] = `31`;

exports[`rollValue: 4d20 3`] = `42`;

exports[`rollValue: 4d20 4`] = `45`;

exports[`rollValue: 4d20 5`] = `37`;
